File: algorithms/1409.queries-on-a-permutation-with-key/solution.rs
impl Solution {
    pub fn process_queries(queries: Vec<i32>, m: i32) -> Vec<i32> {
        let mut p = (1..=m).collect::<Vec<i32>>();

        let mut res: Vec<i32> = vec![];
        for n in queries {
            let idx = p.iter().position(|&x| x == n).unwrap();
            res.push(idx as i32);

            p.remove(idx);
            p.insert(0, n);
        }

        res
    }
}

File: algorithms/1409.queries-on-a-permutation-with-key/solution.go
func processQueries(queries []int, m int) []int {
	p := make([]int, m)
	for i := range p {
		p[i] = i + 1
	}

	res := []int{}
	for _, n := range queries {
		idx := indexOf(n, p)
		res = append(res, idx)

		p = append(p[:idx], p[idx+1:]...)
		p = append([]int{n}, p...)
	}

	return res
}

func indexOf(n int, p []int) int {
	for i, x := range p {
		if n == x {
			return i
		}
	}
	return -1
}
